### PR TITLE
Mix archive.install parse file before installing

### PR DESCRIPTION
**Pull request by tsubery**
This allows to ignore file naming convention and install correctly even if file name is changed.
Solves https://github.com/elixir-lang/elixir/issues/5882

### CODE REVIEW

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 1]
```diff
@@ -84,18 +84,26 @@ defmodule Mix.Tasks.Archive.Install do
...
-    dir_dst = Path.rootname(ez_dst)
-    File.mkdir_p!(dir_dst)
-    {:ok, _} = :zip.extract(contents, [cwd: dir_dst])
-    Mix.shell.info [:green, "* creating ", :reset, Path.relative_to_cwd(dir_dst)]
-
-    ebin = Mix.Local.archive_ebin(dir_dst)
-    Mix.Local.check_elixir_version_in_ebin(ebin)
-    true = Code.append_path(ebin)
-    :ok
+    with {:ok, [_comment, zip_first_file | _]} <- :zip.list_dir(contents),
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: We should probably move all of those checks into a separate function clause and then:

    with {:ok, destination} <- archive_destination(contents)

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 2]
```diff
@@ -84,18 +84,26 @@ defmodule Mix.Tasks.Archive.Install do
...
+
+        remove_previous_versions(previous)
+
+        File.mkdir_p!(dir_dst)
+        {:ok, _} = :zip.extract(contents, [cwd: dir_dst])
+        Mix.shell.info [:green, "* creating ", :reset, Path.relative_to_cwd(dir_dst)]
+
+        ebin = Mix.Local.archive_ebin(dir_dst)
+        Mix.Local.check_elixir_version_in_ebin(ebin)
+      true = Code.append_path(ebin)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Indentation is off.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 3]
```diff
@@ -84,18 +84,26 @@ defmodule Mix.Tasks.Archive.Install do
...
-    File.mkdir_p!(dir_dst)
-    {:ok, _} = :zip.extract(contents, [cwd: dir_dst])
-    Mix.shell.info [:green, "* creating ", :reset, Path.relative_to_cwd(dir_dst)]
-
-    ebin = Mix.Local.archive_ebin(dir_dst)
-    Mix.Local.check_elixir_version_in_ebin(ebin)
-    true = Code.append_path(ebin)
-    :ok
+    with {:ok, [_comment, zip_first_file | _]} <- :zip.list_dir(contents),
+         {:zip_file, zip_first_path, _,_,_,_} <- zip_first_file,
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Please add spaces after commas.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 4]
```diff
@@ -84,18 +84,19 @@ defmodule Mix.Tasks.Archive.Install do
...
-    dir_dst = Path.rootname(ez_dst)
-    File.mkdir_p!(dir_dst)
-    {:ok, _} = :zip.extract(contents, [cwd: dir_dst])
-    Mix.shell.info [:green, "* creating ", :reset, Path.relative_to_cwd(dir_dst)]
-
-    ebin = Mix.Local.archive_ebin(dir_dst)
-    Mix.Local.check_elixir_version_in_ebin(ebin)
-    true = Code.append_path(ebin)
-    :ok
+    with {:ok, dir_dst} <- destination(ez_dst, contents) do
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Is everyone fine with using `dest` instead of `dst`? (And ideally `source` instead of `src`?) :bowtie:

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 5]
```diff
@@ -104,6 +105,18 @@ defmodule Mix.Tasks.Archive.Install do
 
   ### Private helpers
 
+  defp destination(ez_dst, contents) do
+    with {:ok, [_comment, zip_first_file | _]} <- :zip.list_dir(contents),
+         {:zip_file, zip_first_path, _, _, _, _} <- zip_first_file,
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: I assume we are fine to use `=` instead of `<–` here.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 6]
```diff
@@ -104,6 +105,18 @@ defmodule Mix.Tasks.Archive.Install do
...
+         [zip_root_dir | _] <- Path.split(zip_first_path) do
+
+      {:ok, Path.join(Path.dirname(ez_dst), zip_root_dir)}
+    else
+      _ ->
+        {:error, "Invalid archive file"}
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: I think we should not capitalize reason as it will be prefixed later.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 7]
```diff
@@ -104,6 +105,18 @@ defmodule Mix.Tasks.Archive.Install do
 
   ### Private helpers
 
+  defp destination(ez_dst, contents) do
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Let's maybe name it `resolve_destination`.

File: lib/mix/test/mix/tasks/archive_test.exs [hunk 8]
```diff
@@ -70,6 +70,23 @@ defmodule Mix.Tasks.ArchiveTest do
     end
   end
 
+  test "archive install invalid file" do
+    in_fixture "invalid_archive", fn ->
+      assert File.regular? 'bad-archive-0.1.0.ez'
```
Comment (lexmag) [inline lib/mix/test/mix/tasks/archive_test.exs]: Maybe `invalid-archive-0.1.0.ez`?

File: lib/mix/test/mix/tasks/archive_test.exs [hunk 9]
```diff
@@ -70,6 +70,23 @@ defmodule Mix.Tasks.ArchiveTest do
     end
   end
 
+  test "archive install invalid file" do
+    in_fixture "invalid_archive", fn ->
```
Comment (lexmag) [inline lib/mix/test/mix/tasks/archive_test.exs]: I think we can reuse `archive` folder.

File: lib/mix/lib/mix/tasks/archive.install.ex (hunk 6 above)
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: everything else in the module is capitalized so lets keep this as is for now.

Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Messages from other places used verbatim though, let's do the opposite then and move `Installation failed:` here?

Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: We actually raise messages as-is in other places.

Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Yes, let's raise messages as is here too.

File: lib/mix/test/mix/tasks/archive_test.exs (hunk 9 above)
Comment (tsubery) [inline lib/mix/test/mix/tasks/archive_test.exs]: I was concerned building an archive in the same directory as other tests do will cause test pollution but it doesn't seem the case.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 10]
```diff
@@ -104,6 +105,21 @@ defmodule Mix.Tasks.Archive.Install do
...
+  defp resolve_destination(ez_path, contents) do
+    case :zip.list_dir(contents) do
+      {:ok, [_comment, zip_first_file | _]} ->
+        zip_root_dir =
+          zip_first_file
+          |> elem(1)
+          |> Path.split()
+          |> hd
+
+      Path.join(Path.dirname(ez_path), zip_root_dir)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Wrong indentation here.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 11]
```diff
@@ -104,6 +105,21 @@ defmodule Mix.Tasks.Archive.Install do
 
   ### Private helpers
 
+  defp resolve_destination(ez_path, contents) do
+    case :zip.list_dir(contents) do
+      {:ok, [_comment, zip_first_file | _]} ->
+        zip_root_dir =
+          zip_first_file
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Why did you remove the pattern matching? I don't think @lexmag suggested to remove it, rather to use `=` instead of `<-`.

**Comment by josevalim** [comment]
@tsubery thank you, i have added two last comments to a recently refactored function.

File: lib/mix/lib/mix/tasks/archive.install.ex (hunk 10 above)
Comment (tsubery) [inline lib/mix/lib/mix/tasks/archive.install.ex]: fixed

File: lib/mix/lib/mix/local/installer.ex [hunk 12]
```diff
@@ -104,7 +104,7 @@ defmodule Mix.Local.Installer do
     if opts[:force] || should_install?(name, src, previous_files) do
       case Mix.Utils.read_path(src, opts) do
         {:ok, binary} ->
-          module.install(dst, binary, previous_files)
+           module.install(dst, binary, previous_files)
```
Comment (lexmag) [inline lib/mix/lib/mix/local/installer.ex]: Broken indentation.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 13]
```diff
@@ -83,16 +83,17 @@ defmodule Mix.Tasks.Archive.Install do
     end
   end
 
-  def install(ez_dst, contents, previous) do
+  def install(ez_path, contents, previous) do
+    # Get the directory name and extract it there
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: After moving this comment seems misleading, I think we can ditch it.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 14]
```diff
@@ -104,6 +105,18 @@ defmodule Mix.Tasks.Archive.Install do
...
 
+  defp resolve_destination(ez_path, contents) do
+    with {:ok, [_comment, zip_first_file | _]} <- :zip.list_dir(contents),
+         {:zip_file, zip_first_path, _, _, _, _} = zip_first_file,
+         [zip_root_dir | _] = Path.split(zip_first_path) do
+
+        Path.join(Path.dirname(ez_path), zip_root_dir)
+    else
+      _ ->
+        Mix.raise "Installation failed: Invalid archive file"
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/archive.install.ex]: `Invalid` –> `invalid`

File: lib/mix/test/mix/tasks/archive_test.exs [hunk 15]
```diff
@@ -70,6 +70,24 @@ defmodule Mix.Tasks.ArchiveTest do
     end
   end
 
+  test "archive install invalid file" do
+    in_fixture "archive", fn ->
+      file_name = "invalid-archive-0.1.0.ez"
+      assert File.regular?(file_name)
+
+      send self(), {:mix_shell_input, :yes?, true}
+      assert_raise Mix.Error, ~r/Invalid archive file/, fn ->
```
Comment (lexmag) [inline lib/mix/test/mix/tasks/archive_test.exs]: `Invalid` –> `invalid` here as well.

File: lib/mix/lib/mix/tasks/archive.install.ex [hunk 16]
```diff
@@ -104,6 +104,18 @@ defmodule Mix.Tasks.Archive.Install do
 
   ### Private helpers
 
+  defp resolve_destination(ez_path, contents) do
+    with {:ok, [_comment, zip_first_file | _]} <- :zip.list_dir(contents),
+         {:zip_file, zip_first_path, _, _, _, _} = zip_first_file,
+         [zip_root_dir | _] = Path.split(zip_first_path) do
+
+        Path.join(Path.dirname(ez_path), zip_root_dir)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/archive.install.ex]: Wrong indentation. It should be two spaces.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
